src/devices/GoPiGo3/Movements/Vehicle.cs

```diff
@@ -0,0 +1,276 @@
+        }
+
+        /// <summary>
+        /// Stop the vehicule
+        /// </summary>
+        public void Stop()
+        {
+            StopMotor(PortLeft);
+            StopMotor(PortRight);
+        }
+
+        /// <summary>
+        /// Run backward for the specified number of milliseconds
+        /// </summary>
+        /// <param name="speed">speed is between -255 and +255</param>
+        /// <param name="timeout">>number of milliseconds to run the motors</param>
+        public void Backward(int speed, int timeout)
+        {
+            RunMotorSyncTime(new MotorPort[2] { PortLeft, PortRight }, new int[2] { speed * _correctedDir, speed * _correctedDir }, timeout);
+        }
+
+        /// <summary>
+        /// Run forward for the specified number of milliseconds
+        /// </summary>
+        /// <param name="speed">speed is between -255 and +255</param>
+        /// <param name="timeout">>number of milliseconds to run the motors</param>
+        public void Foreward(int speed, int timeout)
+        {
+            Backward(-speed, timeout);
+        }
+
+        /// <summary>
+        /// Return the MotorPort of the left motor 
+        /// </summary>
+        public MotorPort PortLeft { get; }
+
+        /// <summary>
+        /// Return the MotorPort of the right motor
+        /// </summary>
+        public MotorPort PortRight { get; }
+
+        /// <summary>
+        /// Is the vehicule has inverted direction, then true
+        /// </summary>
+        public bool DirectionOpposite
+        {
+            get { return _directionOpposite; }
+
+            set
+            {
+                _directionOpposite = value;
+                _correctedDir = (_directionOpposite) ? -1 : 1;
+            }
+        }
+
+        private void RunMotorSyncTime(MotorPort[] ports, int[] speeds, int timeout)
+        {
+            if ((ports == null) || (speeds == null))
+                return;
+            if (ports.Length != speeds.Length)
+                return;
+            //create a timer for the needed time to run
+            if (_timer == null)
+                _timer = new Timer(RunUntil, ports, TimeSpan.FromMilliseconds(timeout), Timeout.InfiniteTimeSpan);
+            else
+                _timer.Change(TimeSpan.FromMilliseconds(timeout), Timeout.InfiniteTimeSpan);
+
+            //initialize the speed and enable motors
+            for (int i = 0; i < ports.Length; i++)
+            {
+                StartMotor(ports[i], speeds[i]);
+            }
+
+            bool nonstop = true;
+            while (nonstop)
+            {
+                bool status = false;
+                for (int i = 0; i < ports.Length; i++)
+                {
+                    status |= IsRunning(ports[i]);
+                }
+                nonstop = status;
+            }
+        }
+
+        private void RunUntil(object state)
```

Comment:
could this be just bool?

Reply:
If you consider we can trust the Calling function and no modification will be done later on, then , yes, I can simplify. At the end of the day, it's about to stop both motors. Function is calls only for that.